crypto/rand/drbg_lib.c

```diff
@@ -101,8 +106,8 @@ RAND_DRBG *RAND_DRBG_new(int type, unsigned int flags, RAND_DRBG *parent)
  */
 void RAND_DRBG_free(RAND_DRBG *drbg)
 {
-    /* The global DRBG is free'd by rand_cleanup_drbg_int() */
-    if (drbg == NULL || drbg == &rand_drbg)
+    /* The global DRBGs are free'd by rand_drbg_cleanup_int() */
+    if (drbg == NULL || drbg->lock != NULL)
```

Comment:
I'm wondering if we actually ever call this functions with one of the global DRBGs. It's an internal only function, and I don't see the call. If we do call it, we could consider it an error instead, or we could just not check for the global ones and just check drbg == NULL.

Reply:
The original line 105 has been there since 75e2c877650444fb829547bdb58d46eb1297bc1a, but the check for `priv_drbg` was not amended when the private DRBG was introduced in ddc6a5c8f5900959bdbdfee79e1625a3f7808acd. Instead of adding `|| drbg == &priv_drbg`, I used this more general condition, anticipating the presence of a third master DRBG in #4402.

So you probably have to ask @richsalz about his intention. I think the check is there to  prevent the caller from accidentally freeing one of the global DRBGs. Since the RAND_DRBG API might become public in the future, in principle it is not bad to add safety checks early.

However, freeing static memory is a cardinal beginners error and it is not common for C libraries to check against stupid errors. A standard library function like free() would just happily crash the program and this would be its just punishment. So I would go with your last suggestion to keep only the NULL pointer check.

Reply:
By the way, the name `drbg_free()` and the comment preceding the function are extremely misleading, since the DRBG itself is *not* freed.
```C
/* Clean up a DRBG and free it */
static void drbg_free(RAND_DRBG *drbg)
{
    CRYPTO_THREAD_lock_free(drbg->lock);
    RAND_DRBG_uninstantiate(drbg);
}
```
I will change the name to `drbg_cleanup()`, which is more appropriate for the antagonist to `drbg_setup()`:
```C
/* Cleans up the given global DRBG  */
static void drbg_cleanup(RAND_DRBG *drbg)
{
    CRYPTO_THREAD_lock_free(drbg->lock);
    RAND_DRBG_uninstantiate(drbg);
}
```

Reply:
The original check was there because I wasn't sure all code-paths did only the right thing.  I'm happy to remove it :)